File: src/test/scala/tlschannel/BlockingTest.scala
package tlschannel

import org.junit.jupiter.api.{DynamicTest, TestFactory, TestInstance}
import org.junit.jupiter.api.TestInstance.Lifecycle
import tlschannel.helpers.TestUtil.LazyListWithTakeWhileInclusive
import tlschannel.helpers.SslContextFactory
import tlschannel.helpers.SocketPairFactory
import tlschannel.helpers.Loops
import tlschannel.helpers.SocketPairFactory.ChunkSizeConfig
import tlschannel.helpers.SocketPairFactory.ChuckSizes

import java.util
import java.util.logging.Logger
import scala.jdk.CollectionConverters._

@TestInstance(Lifecycle.PER_CLASS)
class BlockingTest {

  val logger = Logger.getLogger(classOf[BlockingTest].getName)

  val sslContextFactory = new SslContextFactory

  val factory = new SocketPairFactory(sslContextFactory.defaultContext)
  val dataSize = 60 * 1000

  // Test a half-duplex interaction, with renegotiation before reversing the direction of the flow (as in HTTP)
  @TestFactory
  def testHalfDuplexWireRenegotiations(): util.Collection[DynamicTest] = {
    println("testHalfDuplexWireRenegotiations():")
    val sizes = LazyList.iterate(1)(_ * 3).takeWhileInclusive(_ <= SslContextFactory.tlsMaxDataSize)
    val tests = for ((size1, size2) <- sizes zip sizes.reverse) yield {
      DynamicTest.dynamicTest(
        s"testHalfDuplexWireRenegotiations() - size1=$size1, size2=$size2",
        () => {
          val socketPair = factory.nioNio(
            chunkSizeConfig = Some(
              ChunkSizeConfig(
                ChuckSizes(Some(size1), Some(size2)),
                ChuckSizes(Some(size1), Some(size2))
              )
            )
          )
          Loops.halfDuplex(socketPair, dataSize, renegotiation = true)
          println(f"$size1%5d -eng-> $size2%5d -net-> $size1%5d -eng-> $size2%5d")
        }
      )
    }
    tests.asJava
  }

  // Test a full-duplex interaction, without any renegotiation
  @TestFactory
  def testFullDuplex(): util.Collection[DynamicTest] = {
    println("testFullDuplex():")
    val sizes = LazyList.iterate(1)(_ * 3).takeWhileInclusive(_ <= SslContextFactory.tlsMaxDataSize)
    val tests = for ((size1, size2) <- sizes zip sizes.reverse) yield {
      DynamicTest.dynamicTest(
        s"testFullDuplex() - size1=$size1,size2=$size2",
        () => {
          logger.fine(() => s"Testing sizes: size1=$size1,size2=$size2")
          val socketPair = factory.nioNio(chunkSizeConfig =
            Some(
              ChunkSizeConfig(
                ChuckSizes(Some(size1), Some(size2)),
                ChuckSizes(Some(size1), Some(size2))
              )
            )
          )
          Loops.fullDuplex(socketPair, dataSize)
          println(f"$size1%5d -eng-> $size2%5d -net-> $size1%5d -eng-> $size2%5d")
        }
      )
    }
    tests.asJava
  }

}

File: src/test/scala/tlschannel/BlockingTest.java
package tlschannel;

import java.util.*;
import java.util.stream.Collectors;
import org.junit.jupiter.api.DynamicTest;
import org.junit.jupiter.api.TestFactory;
import org.junit.jupiter.api.TestInstance;
import org.junit.jupiter.api.TestInstance.Lifecycle;
import scala.Option;
import tlschannel.helpers.Loops;
import tlschannel.helpers.SocketGroups.SocketPair;
import tlschannel.helpers.SocketPairFactory;
import tlschannel.helpers.SocketPairFactory.ChuckSizes;
import tlschannel.helpers.SocketPairFactory.ChunkSizeConfig;
import tlschannel.helpers.SslContextFactory;
import tlschannel.util.ListUtils;
import tlschannel.util.StreamUtils;

@TestInstance(Lifecycle.PER_CLASS)
public class BlockingTest {

    private static final int dataSize = 60 * 1000;
    private final SslContextFactory sslContextFactory = new SslContextFactory();
    private final SocketPairFactory factory = new SocketPairFactory(sslContextFactory.defaultContext());

    // Test a half-duplex interaction, with renegotiation before reversing the direction of the flow (as in HTTP)
    @TestFactory
    public Collection<DynamicTest> testHalfDuplexWireRenegotiations() {
        System.out.println("testHalfDuplexWireRenegotiations():");
        List<Integer> sizes = StreamUtils.iterate(1, x -> x < SslContextFactory.tlsMaxDataSize * 2, x -> x * 2)
                .collect(Collectors.toList());
        List<Integer> reversedSizes = ListUtils.reversed(sizes);
        List<DynamicTest> ret = new ArrayList<>();
        for (int i = 0; i < sizes.size(); i++) {
            int size1 = sizes.get(i);
            int size2 = reversedSizes.get(i);
            ret.add(DynamicTest.dynamicTest(
                    String.format("testHalfDuplexWireRenegotiations() - size1=%d, size2=%d", size1, size2), () -> {
                        SocketPair socketPair = factory.nioNio(
                                Option.apply(null),
                                Option.apply(new ChunkSizeConfig(
                                        new ChuckSizes(Optional.of(size1), Optional.of(size2)),
                                        new ChuckSizes(Optional.of(size1), Optional.of(size2)))),
                                true,
                                false,
                                Option.apply(null));
                        Loops.halfDuplex(socketPair, dataSize, true, false);
                        System.out.printf("%5d -eng-> %5d -net-> %5d -eng-> %5d\n", size1, size2, size1, size2);
                    }));
        }
        return ret;
    }

    // Test a full-duplex interaction, without any renegotiation
    @TestFactory
    public Collection<DynamicTest> testFullDuplex() {
        System.out.println("testFullDuplex():");
        List<Integer> sizes = StreamUtils.iterate(1, x -> x < SslContextFactory.tlsMaxDataSize * 2, x -> x * 2)
                .collect(Collectors.toList());
        List<Integer> reversedSizes = ListUtils.reversed(sizes);
        List<DynamicTest> ret = new ArrayList<>();
        for (int i = 0; i < sizes.size(); i++) {
            int size1 = sizes.get(i);
            int size2 = reversedSizes.get(i);
            ret.add(DynamicTest.dynamicTest(
                    String.format("testFullDuplex() - size1=%d, size2=%d", size1, size2), () -> {
                        SocketPair socketPair = factory.nioNio(
                                Option.apply(null),
                                Option.apply(new ChunkSizeConfig(
                                        new ChuckSizes(Optional.of(size1), Optional.of(size2)),
                                        new ChuckSizes(Optional.of(size1), Optional.of(size2)))),
                                true,
                                false,
                                Option.apply(null));
                        Loops.fullDuplex(socketPair, dataSize);
                        System.out.printf("%5d -eng-> %5d -net-> %5d -eng-> %5d\n", size1, size2, size1, size2);
                    }));
        }
        return ret;
    }
}
